### ISSUE TITLE
All decision node are not parsed before raising exception

### ISSUE THREAD

**Issue by mumbel** [NationalSecurityAgency/ghidra#2097]
master b3b7bab4cade9bbef17d377117ba7e228e0a79b6

Was looking at an ARM:LE:32:v8 elf (also verified with just a 4 byte binary), which has a `e6b25015 	ssat	r5, #19, r5`  

Looking at DebugSleighInstructionParse, Ghidra quickly narrows it down to 4 instructions:
```
      : {line# 293} <instruction>
      : {line# 294} <instruction>
      : {line# 2867} ldrt<COND> <Rd>,<addrmode2>
      : {line# 5230} ssat<COND> <Rd>, <sSatImm5>, <shift4>
```

The script fails the first two `:^instruction`, but decides `ldrt` is a match of the byte and context pattern.  

It begins to parse the `COND`, `Rd`, and `addrmode2`.  `addrmode2` cannot be resolved due to `addr2shift` not resolving, due to `c0404=0` failing.

An exception is raised and `15 50 b2 e6` is marked `Unable to resolve constructor`.  

If I move `ssat` above `ldrt` in ARMinstructions.sinc, it does disassemble correctly.  Also tried putting all the addr2shift into a `with : c0404=0` (changed c0411=0 to c0511=0), wondering if it would maybe catch the 4th bit set in the byte pattern, no luck.

traced the match down to:
https://github.com/NationalSecurityAgency/ghidra/blob/49c2010b63b56c8f20845f3970fedd95d003b1e9/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/app/plugin/processors/sleigh/DecisionNode.java#L70

Parses index 0, 1, and 2 of 4 elements.  Can this come back here after the fact of matching/failing on index 2 and try index 3?

**Comment by GhidorahRex** [comment]
This is a known issue with how sleigh resolves constructors. Once it has decided on a constructor match, it gets locked in, even if a sub-constructor fails.

The fix for this situation will probably be moving the `ssat` constructor above the `ldrt` constructor, as you tested. There are several other constructor patterns in ARM that have this requirement.

**Comment by ghidra1** [comment]
Unfortunately, the Ghidra ARM instruction patterns are not always mutually exclusive and rely heavily on ordering which we generally try to avoid since it can be very brittle and hidden. We generally encourage constructors be mutually exclusive with their pattern match.  Use of the sleigh -l (lowercase L) option is used to enforce this during language compilation.  You can note that attempting to use this option with ARM fails badly.  ARM seems to leave a lot of holes in their instruction patterns and use those for other instructions.  Unfortunately, it is almost impossible to verify that adding a new instruction will not break another without exercising every instruction parse - which we don't do very well.

**Comment by mumbel** [comment]
Thanks for the reply @ghidra1 and @GhidorahRex 

I didn't think to look at the `sleigh` output to verify, thanks.

I think I remember coming across this now when I was doing some x86, but figured that was just due to variable length instructions and prefix parsing (which I guess could still hit fixed length or anything with a context).  Please close this issue if having it open isn't needed anymore, but I'll leave open for future reference incase.

**Comment by ghidra1** [comment]
We still need to address the instruction parse issue and should probably leave this ticket open until it is resolved.  The issue could be more widespread if we are failing to parse other instructions for a similar reason.  In the case of moving ssat (line 5230) to above ldrt (line 2867) we would need to verify that ssat still parses properly.  We may also want to determine what part of the pattern causes this and identify other instructions with the same sensitivity.

**Comment by ghidra1** [comment]
Oops ... need to revise my comment above.  We should only move instructions down within the file, no up.  Moving instructions up could possible negatively affect parsing of other instructions.

**Comment by GhidorahRex** [comment]
After reviewing the issue, it looks like the ldrt constructor is missing the (I25=0 | (I25=1 & c0404=0)) piece that other addrmode2 instructions include. Adding that to the constructor also corrects the disassembly error.

**Comment by mumbel** [comment]
looks closed by 558844aaa9af40d5f2410d25ff92e34f5188292a
